internal/lang/funcs/string.go

```diff
@@ -164,3 +166,110 @@ func Replace(str, substr, replace cty.Value) (cty.Value, error) {
 func StrContains(str, substr cty.Value) (cty.Value, error) {
 	return StrContainsFunc.Call([]cty.Value{str, substr})
 }
+
+// MakeTemplateStringFunc constructs a function that takes a string and
+// an arbitrary object of named values and attempts to render that string
+// as a template using HCL template syntax.
+func MakeTemplateStringFunc(content string, funcsCb func() map[string]function.Function) function.Function {
+
+	params := []function.Parameter{
+		{
+			Name:        "data",
+			Type:        cty.String,
+			AllowMarked: true,
+		},
+		{
+			Name:        "vars",
+			Type:        cty.DynamicPseudoType,
+			AllowMarked: true,
+		},
+	}
+	loadTmpl := func(content string, marks cty.ValueMarks) (hcl.Expression, error) {
+
+		expr, diags := hclsyntax.ParseTemplate([]byte(content), "NoFileNeeded", hcl.Pos{Line: 1, Column: 1})
+		if diags.HasErrors() {
+			return nil, diags
+		}
+
+		return expr, nil
+	}
+
+	renderTmpl := func(expr hcl.Expression, varsVal cty.Value) (cty.Value, error) {
+		if varsTy := varsVal.Type(); !(varsTy.IsMapType() || varsTy.IsObjectType()) {
+			return cty.DynamicVal, function.NewArgErrorf(1, "invalid vars value: must be a map") // or an object, but we don't strongly distinguish these most of the time
+		}
+
+		ctx := &hcl.EvalContext{
+			Variables: varsVal.AsValueMap(),
+		}
+
+		// We require all of the variables to be valid HCL identifiers, because
+		// otherwise there would be no way to refer to them in the template
+		// anyway. Rejecting this here gives better feedback to the user
+		// than a syntax error somewhere in the template itself.
+		for n := range ctx.Variables {
+			if !hclsyntax.ValidIdentifier(n) {
+				// This error message intentionally doesn't describe _all_ of
+				// the different permutations that are technically valid as an
+				// HCL identifier, but rather focuses on what we might
+				// consider to be an "idiomatic" variable name.
+				return cty.DynamicVal, function.NewArgErrorf(1, "invalid template variable name %q: must start with a letter, followed by zero or more letters, digits, and underscores", n)
+			}
+		}
+
+		// We'll pre-check references in the template here so we can give a
+		// more specialized error message than HCL would by default, so it's
+		// clearer that this problem is coming from a templatestring call.
+		for _, traversal := range expr.Variables() {
+			root := traversal.RootName()
+			if _, ok := ctx.Variables[root]; !ok {
+				return cty.DynamicVal, function.NewArgErrorf(1, "vars map does not contain key %q", root)
+			}
+		}
+
+		givenFuncs := funcsCb() // this callback indirection is to avoid chicken/egg problems
+		funcs := make(map[string]function.Function, len(givenFuncs))
+		for name, fn := range givenFuncs {
+			funcs[name] = fn
+		}
+		ctx.Functions = funcs
+
+		val, diags := expr.Value(ctx)
+		if diags.HasErrors() {
+			return cty.DynamicVal, diags
+		}
+		return val, nil
+	}
```

Comment:
This function seems to be copied verbatim from `filesystem.go`. It serves the same purpose, so there is no reason to duplicate this function.  This needs to be refactored so both the templatefile and templatestring function can use the same function instead of copying, otherwise the code will have to be maintained in two places, which is less than ideal.

Reply:
At start of the PR, I had used the same piece of code for both the functions with the respective modifications needed but then @Yantrio raised concerns regarding potential confusion for end-users who would look for the templatestring function, they would be confused to find it in filesystem.go and as templatestring is related to string so it was suggested that its logic be moved to the string.go file to ensure clarity so that's why I created new function copying all logic into string.go file for templatestring

Reply:
@sanskruti-shahu if this is the exact same function, it should definitely not be copied over. Instead, this specific function should be moved out from both here and the templatefile function and called accordingly. @Yantrio was right, the templatefile function should not use the templatestring function, but copying 50-ish lines of code is also not the right way to go.

Reply:
@janosdebugs I got your point but I am not clear about what am I supposed to do next? Do I have to make a third function which will contain the logic which is common in both and it will be called separately for templatefile and templatestring and if this is so then where shall I place that function? creating new file or inside any (don't know which) existing file?

Reply:
Yes, the renderTmpl function should be extracted into possibly a separate file.

Reply:
Ok then I'll create new file having func containing the logic of renderTmpl which will be called by both templatestring and templatefile.